pkg/daemon/imagepuller/puller.go

```diff
@@ -0,0 +1,457 @@
+	w.Lock()
+	defer w.Unlock()
+	if w.active {
+		for _, worker := range w.pullWorkers {
+			worker.Stop()
+		}
+		w.active = false
+	}
+}
+
+func (w *realWorkerPool) UpdateStatus(status *appsv1alpha1.ImageTagStatus) {
+	w.Lock()
+	defer w.Unlock()
+	if !w.active {
+		return
+	}
+	w.tagStatuses[status.Tag] = status
+}
+
+func newPullWorker(name string, tagSpec appsv1alpha1.ImageTagSpec, secrets []v1.Secret, runtime daemonruntime.ImageRuntime, statusUpdater imageStatusUpdater, ref *v1.ObjectReference, eventRecorder record.EventRecorder) *pullWorker {
+	o := &pullWorker{
+		name:          name,
+		tagSpec:       tagSpec,
+		secrets:       secrets,
+		runtime:       runtime,
+		statusUpdater: statusUpdater,
+		ref:           ref,
+		eventRecorder: eventRecorder,
+		active:        true,
+		stopCh:        make(chan struct{}),
+	}
+	go o.Run()
+	return o
+}
+
+type pullWorker struct {
+	sync.Mutex
+
+	name          string
+	tagSpec       appsv1alpha1.ImageTagSpec
+	secrets       []v1.Secret
+	runtime       daemonruntime.ImageRuntime
+	statusUpdater imageStatusUpdater
+	ref           *v1.ObjectReference
+	eventRecorder record.EventRecorder
+
+	active bool
+	stopCh chan struct{}
+}
+
+func (w *pullWorker) ImageRef() string {
+	return fmt.Sprintf("%v:%v", w.name, w.tagSpec.Tag)
+}
+
+func (w *pullWorker) Stop() {
+	w.Lock()
+	defer w.Unlock()
+	if w.active {
+		klog.Warningf("Worker to pull image %s:%s is stopped", w.name, w.tagSpec.Tag)
+		close(w.stopCh)
+		w.active = false
+	}
+}
+
+func (w *pullWorker) IsActive() bool {
+	return w.active
+}
+
+func (w *pullWorker) Run() {
+	if !w.active {
+		return
+	}
+
+	klog.V(3).Infof("starting worker %v version %v", w.ImageRef(), w.tagSpec.Version)
+
+	tag := w.tagSpec.Tag
+	startTime := metav1.Now()
+	newStatus := &appsv1alpha1.ImageTagStatus{
+		Tag:       tag,
+		Phase:     appsv1alpha1.ImagePhasePulling,
+		StartTime: &startTime,
+		Version:   w.tagSpec.Version,
+	}
+	defer func() {
+		cost := time.Now().Sub(startTime.Time)
+		if newStatus.Phase == appsv1alpha1.ImagePhaseFailed {
+			klog.Warningf("Worker failed to pull image %s:%s, cost %v, err: %v", w.name, tag, cost, newStatus.Message)
+		} else {
+			klog.Infof("Successfully pull image %s:%s, cost %vs", w.name, tag, cost)
+		}
+		w.statusUpdater.UpdateStatus(newStatus)
+	}()
+
+	timeout := defaultImagePullingTimeout
+	if w.tagSpec.PullPolicy != nil && w.tagSpec.PullPolicy.TimeoutSeconds != nil {
+		timeout = time.Duration(*w.tagSpec.PullPolicy.TimeoutSeconds) * time.Second
+	}
+	backoffLimit := defaultImagePullingBackoffLimit
+	if w.tagSpec.PullPolicy != nil && w.tagSpec.PullPolicy.BackoffLimit != nil {
+		backoffLimit = int(*w.tagSpec.PullPolicy.BackoffLimit)
+	}
+
+	var (
+		step       = time.Second
+		maxBackoff = 30 * time.Second
+	)
+
+	var lastError error
+	for i := 0; i < backoffLimit; i++ {
+		pullContext, cancel := context.WithTimeout(context.Background(), timeout)
+		lastError = w.doPullImage(pullContext, newStatus)
+		if lastError == nil {
```

Comment:
it looks more clear if we check lastError != nil and do the error handling here, and leave the normal handling outside if branch